### ISSUE TITLE
os: file reads do not detect end-of-file

### ISSUE THREAD

**Issue by leighmcculloch** [vlang/v#10036]
**V version:** 0.2.2 d6a963a
**OS:** Ubuntu Focal 20.04

**What did you do?**
```v
module main

import io
import os

fn main() {
	io.cp(os.stdout(), os.stdin()) ?
}
```
```sh
$ echo hi | v run main.v
```

**What did you expect to see?**
I expected to see the program print `hi`, then exit.

```sh
$ echo hi | v run main.v
hi

$
```
 
**What did you see instead?**
The program prints `hi`, then blocks waiting for more input even though the input stream has finished.

```sh
$ echo hi | v run main.v
hi
```

**More info**
As a comparison, the same program written in Go will exit.

```go
package main

import (
	"io"
	"os"
)

func main() {
	_, err := io.Copy(os.Stdout, os.Stdin)
	if err != nil {
		panic(err)
	}
}
```
```sh
$ echo hi | go run main.go
hi

$
```

**Comment by leighmcculloch** [comment]
On closer inspection it appears that there is a discrepancy in different parts of the code about what it means for the end-of-file to have been reached.

For example:

- The `io.cp()` function exits only if the reader returns an error:
   https://github.com/vlang/v/blob/43cf3542ac06d5c09ea28633ceb2b9868a27805d/vlib/io/io.v#L7-L10

- The `io.cp()` tests appear to back that expectation up:
   https://github.com/vlang/v/blob/43cf3542ac06d5c09ea28633ceb2b9868a27805d/vlib/io/io_test.v#L15-L18

- However, when reading from a file, or from stdin, and the end-of-file is reached, then no error is returned and instead a `0` bytes read value is returned. The code here appears to back that up:
   https://github.com/vlang/v/blob/43cf3542ac06d5c09ea28633ceb2b9868a27805d/vlib/os/file.c.v#L361-L372

I think what's missing is that the file reading process should be checking the return value of `C.feof()` after each `C.fread`. Without doing that I don't think the file reading code will ever return a `none` to signal the end-of-file.

If this looks like the right approach, I can open a PR fixing it.

**Comment by leighmcculloch** [comment]
Another note, the file reading code checks `errno`, but `fread`, as far as I know, will not set `errno`. `read` will set `errno`, but not `fread`.

**Comment by bstnbuck** [comment]
`fread` does not set `errno`, that is correct. But all `fseek` functions do.

**Comment by bstnbuck** [comment]
To fix that the return value of `len` must be checked if it is 0 or negative. After that it works like in Go. I think `os.read()` and so on are working as excepted. I will fix it soon.

**Comment by leighmcculloch** [comment]
> To fix that the return value of `len` must be checked if it is 0 or negative.

I think we need to use `ferror()` and feof()`. All the man docs I've read have said that using those functions are the only way to distinguish between and eof and some other error. Ideally we return the `none` error from read if eof.

**Comment by leighmcculloch** [comment]
> `fread` does not set `errno`, that is correct. But all `fseek` functions do.

In quite a few places in file.c.v the fseek function is called, then errno is set to zero, then fread is called, then errno is checked. It seems to me like we are discarding the errno values for fseek and then trying to use it for fread. An I misunderstanding how it works?

For example:

https://github.com/vlang/v/blob/43cf3542ac06d5c09ea28633ceb2b9868a27805d/vlib/os/file.c.v#L361-L372

**Comment by bstnbuck** [comment]
> 
> 
> > To fix that the return value of `len` must be checked if it is 0 or negative.
> 
> What return value will indicate eof?
> 
> I think we need to use `ferror()` and `feof()`. All the man docs I've read have said that using those functions are the only way to distinguish between and eof and some other error. Ideally we return the `none` error from read if eof.
https://github.com/vlang/v/blob/d26ac5692e174967d4c9caae2bf4d62f2fbe76e1/vlib/io/io.v#L7-L14

If nothing was read, os.read() returns 0 not `none`. 
If you use `feof()` with `os.read()` it always will return an error, because the eof is still at the end of a file. But I think `ferror()` could be useful. This is my solution for `io.cp()` to get the same behaviour like in Go:
```v
pub fn cp(dst Writer, src Reader) ? {
	mut buf := []byte{len: buf_max_len}
	for {
		len := src.read(mut buf) or { break }
		if len <= 0{
			break
		}
		dst.write(buf[..len]) or {
			return err
		}
	}
	unsafe {
		buf.free()
	}
}
```
What do you think?

**Comment by bstnbuck** [comment]
> 
> 
> > `fread` does not set `errno`, that is correct. But all `fseek` functions do.
> 
> In quite a few places in file.c.v the fseek function is called, then errno is set to zero, then fread is called, then errno is checked. It seems to me like we are discarding the errno values for fseek and then trying to use it for fread. An I misunderstanding how it works?
> 
> For example:
> 
> https://github.com/vlang/v/blob/43cf3542ac06d5c09ea28633ceb2b9868a27805d/vlib/os/file.c.v#L361-L372

Yep, I meant that it should be, not that it is. ;) Of course, this should be checked for `fseek()`, not for `fread()`. The question is whether such error checking should only happen at the end before the return or after each `fseek`. What do you think about this?
Something like this?
```v
pub fn (f &File) read(mut buf []byte) ?int {
	if buf.len == 0 {
		return 0
	}
	nbytes := int(C.fread(buf.data, 1, buf.len, f.cfile))
	if C.ferror(f.cfile) != 0 {
		return error("fread failed!")
	}
	return nbytes
}
```
```v
// os.read_bytes_into()...
$if x64 {
		$if windows {
			// Note: fseek errors if pos == os.file_size, which we accept
			C.errno = 0
			C._fseeki64(f.cfile, pos, C.SEEK_SET)
			// errno is only set if fread fails, so clear it first to tell
			if C.errno != 0 {
				return error(posix_get_error_msg(C.errno))
			}
			nbytes := int(C.fread(buf.data, 1, buf.len, f.cfile))
			if C.ferror(f.cfile) != 0 {
				return error("fread failed!")
			}
```

**Comment by leighmcculloch** [comment]
Hmm, I'm not sure now, I think I need to do some tests. I'll have a look soon.

**Comment by leighmcculloch** [comment]
I still need to do some tests to confirm but I think we need to use feof() and if it indicates eof return none, then if ferror return error.Are we hoping to get posix errors for read? We might need to use the read function instead of fread to get posix errors.

**Comment by leighmcculloch** [comment]
Okay, after some testing this is what I think is required for simply focusing on fixing the fread usage. I think we should focus on fixing it in isolation and fix the seek error detection separately.

This diff fixes just the `read` function, but the same pattern can be applied to all other fread usages I think.

We should also add tests that read to the end of a file. None of the tests do that by the looks of it.

```diff
diff --git a/vlib/os/file.c.v b/vlib/os/file.c.v
index cdc612db7..cb969cc67 100644
--- a/vlib/os/file.c.v
+++ b/vlib/os/file.c.v
@@ -391,10 +391,19 @@ pub fn (f &File) read(mut buf []byte) ?int {
        if buf.len == 0 {
                return 0
        }
-       C.errno = 0
        nbytes := int(C.fread(buf.data, 1, buf.len, f.cfile))
-       if C.errno != 0 {
-               return error(posix_get_error_msg(C.errno))
+       // If fread encounters an error, return it.
+       if C.ferror(f.cfile) != 0 {
+               return error("read failed: $f.cfile")
+       }
+       // If fread encounters end-of-file, and no data was read, return the none
+       // error. Note that fread may read data and encounter the end-of-file, but
+       // we shouldn't return none in that case because data was written to buf and
+       // we should inform the caller of that. The caller will get none on their
+       // next call because there will be no data available and the end-of-file
+       // will be encountered again.
+       if nbytes == 0 && C.feof(f.cfile) != 0 {
+               return none
        }
        return nbytes
 }
```

>I will fix it soon.

@bstnbuck Are you volunteering to take this issue? Thanks! 🎉

**Comment by bstnbuck** [comment]
> 
> 
> I still need to do some tests to confirm but I think we need to use feof() and if it indicates eof return none, then if ferror return error.Are we hoping to get posix errors for read? We might need to use the read function instead of fread to get posix errors.

I don't think using `read()` instead of `fread()` is a good solution. The reason is that `fread()` does buffering, but `read()` does not. In addition, only very few errors can occur because a file must already be open.

**Comment by bstnbuck** [comment]
> 
> 
> Okay, after some testing this is what I think is required for simply focusing on fixing the fread usage. I think we should focus on fixing it in isolation and fix the seek error detection separately.
> 
> This diff fixes just the `read` function, but the same pattern can be applied to all other fread usages I think.
> 
> We should also add tests that read to the end of a file. None of the tests do that by the looks of it.
> 
> It'd also be ideal if the file path was stored in the `File` struct so that we could include the filename in the read error. Not sure if there's a tradeoff there. Probably should be out-of-scope for this issue.
> 
> What do you think?
> 
> ```diff
> diff --git a/vlib/os/file.c.v b/vlib/os/file.c.v
> index cdc612db7..cb969cc67 100644
> --- a/vlib/os/file.c.v
> +++ b/vlib/os/file.c.v
> @@ -391,10 +391,19 @@ pub fn (f &File) read(mut buf []byte) ?int {
>         if buf.len == 0 {
>                 return 0
>         }
> -       C.errno = 0
>         nbytes := int(C.fread(buf.data, 1, buf.len, f.cfile))
> -       if C.errno != 0 {
> -               return error(posix_get_error_msg(C.errno))
> +       // If fread encounters an error, return it.
> +       if C.ferror(f.cfile) != 0 {
> +               return error("read failed: $f.cfile")
> +       }
> +       // If fread encounters end-of-file, and no data was read, return the none
> +       // error. Note that fread may read data and encounter the end-of-file, but
> +       // we shouldn't return none in that case because data was written to buf and
> +       // we should inform the caller of that. The caller will get none on their
> +       // next call because there will be no data available and the end-of-file
> +       // will be encountered again.
> +       if nbytes == 0 && C.feof(f.cfile) != 0 {
> +               return none
>         }
>         return nbytes
>  }
> ```
> 
> > I will fix it soon.
> 
> @bstnbuck Are you volunteering to take this issue? I'm happy to take it, but don't want to if you're keen.

Very good work! This solution should be the state of the art. 

Feel free to take on the bug fixing. 

I would also include the file name in the error. What could such a test for EOF look like? 

However, please fix other `fread()` errors in the following functions. Also, `fwrite()` should at least be queried with `ferror()`. 
* file.c.v
  * os.write(), os.writeln(), os.write_string(), os.write_to(), os.write_ptr(), os.write_ptr_at()
  * os.read_bytes_into(), os.read(), os.read_from(), 
  * os.read_struct(), os.read_struct_at(), os.read_raw(), os.read_raw_at(), 
  * os.write_struct(), os.write_struct_at(), os.write_raw(), os.write_raw_at(), 
* os_c.v
  * os.read_bytes(), os.read_file(),

**Comment by bstnbuck** [comment]
If you want, you can also fix the `seek()` errors (preferably in a separate pull). What do you think of this?
```v
if C.fseek(f.cfile, pos, C.SEEK_SET) != 0 {
	return error(posix_get_error_msg(C.errno))
}
```

The following functions must be fixed:
* file.c.v
  * os.write_to(), os.write_ptr_at(), os.write_struct_at(), os.write_raw_at()
  * os.read_bytes_into(), os.read_from(), os.read_struct_at(), os.read_raw_at()
* os_c.v
  * os.read_bytes(), os.read_file(), os.read_file_array(),

**Comment by leighmcculloch** [comment]
I opened a PR fixing the bug: #10070.

However, the tests highlighted that there is one test that expects end-of-file to result in no error, rather than return `none`:
https://github.com/vlang/v/blob/ebe58dcafa40cc1d4bed147631ef0dd8d2921c3e/vlib/os/os_test.v#L160-L166

The test was written after the file code was written so there's a good chance the test was written this way to capture existing behavior.

I've updated the PR changing this test, but I'm not sure if that is the right thing to do.

Probably need some feedback here from folks who have worked on these files significantly. I've picked some names from the git history: @spytheman @ntrel @emily33901 @medvednikov.
